test/Microsoft.ML.Tests/Transformers/KeyToVectorEstimatorTests.cs

```diff
@@ -222,5 +223,51 @@ public void TestOldSavingAndLoading()
                 var loadedView = ModelFileUtils.LoadTransforms(Env, dataView, ms);
             }
         }
+
+        public class ModelInput
+        {
+            [ColumnName("Label"), LoadColumn(0)]
+            public int Label { get; set; }
+
+
+            [ColumnName("ProblematicColumn"), LoadColumn(1)]
+            public string ProblematicColumn { get; set; }
+        }
+
+        static IEnumerable<ModelInput> GetData()
+        {
+            for (int i = 0; i < 1000; i++)
+            {
```

Comment:
Are we seeing overflow errors at only 1000 unique values? Does this mean we cannot perform one hot hash encoding on a 1000 unique values?

Reply:
In this case, yes. Can't we perform OneHotHashEncoding on a thousand values? I don't think so.
This pipeline is to simulate how AutoML reacts toward a bigger dataset using a small dataset.
If we merely use OneHotHashEncoding for a thousand unique values there should be no problem(no overflow). However, the pipeline designed by AutoML appends a ColumnCopy pipeline after OneHotHashEncoding which causes the overflow. But normally(My view) we do not do this way.

Reply:
Does the bug repro if we don't have a CopyColumns after OneHotHashEncoding?


---
In reply to: [439639201](https://github.com/dotnet/machinelearning/pull/5232#discussion_r439639201) [](ancestors = 439639201)

Reply:
no

Reply:
That sounds a little suspicious. Can you please dig a little deeper? CopyColumns shouldn't be impacting the performance of OneHotHashEncoding.

---
In reply to: [439641032](https://github.com/dotnet/machinelearning/pull/5232#discussion_r439641032) [](ancestors = 439641032)

Reply:
sure

Reply:
Wait... I'm wrong on your last question. We still can repo the overflow even if without CopyColumns(). So the CopyColumn() is not the reason here. 
Right now the I see the difference is that in AutoML the OneHothashEncoder has been used twice which means:
in normal case:
var result = pipeline.Fit(dataView).Transform(dataView);
in AutoML:
var transformer= pipeline2.Fit(pipeline.Fit(dataView).Transform(dataView)) (pipeline2 and pipeline both have oneHotHashEncoder)

Reply:
Can you please fix the pipeline to the specific conditions that will reproduce the issue? (If CopyColumns is not necessary to reproduce the issue, that shouldn't be part of the pipeline)

---
In reply to: [439646321](https://github.com/dotnet/machinelearning/pull/5232#discussion_r439646321) [](ancestors = 439646321)

Reply:
It's fixed

Reply:
Sorry, I think I was looking at an older iteration. 
This looks good for the ML.NET portion. Do you know why the AutoML.NET code is calling Fit twice?

---
In reply to: [440403606](https://github.com/dotnet/machinelearning/pull/5232#discussion_r440403606) [](ancestors = 440403606)

Reply:
> Sorry, I think I was looking at an older iteration.
> This looks good for the ML.NET portion. Do you know why the AutoML.NET code is calling Fit twice?
> 
> In reply to: [440403606](https://github.com/dotnet/machinelearning/pull/5232#discussion_r440403606) [](ancestors = 440403606)

My understanding is that the first step AutoML does is to find the best way to featurize the original data(this step calls fit and transform), and then feed the model using the featurized data(this step calls fit again)

Reply:
Why does calling Fit twice cause an overflow?

---
In reply to: [440407700](https://github.com/dotnet/machinelearning/pull/5232#discussion_r440407700) [](ancestors = 440407700)

Reply:
As in the second fit(), a Vector<Key<UInt32, 0-65535>, 65536> is required since OneHotHashTransformer is used twice. And that requires a size of 65536*65536 unit of memory which causes overflow

Reply:
The bug occurs not only when you call Fit twice but also if you have two calls to `OneHotHashEncoding` twice in the same pipeline with the same column name.  The behavior of ML.NET is correct in this case.  
However there may be a bug in AutoML.NET in this case because calling it squares the size of the one hot encoded vector and this may be something that is unintended. 

Does AutoML.NET generate code that has the `Fit()` being called twice?





---
In reply to: [440407700](https://github.com/dotnet/machinelearning/pull/5232#discussion_r440407700) [](ancestors = 440407700)

Reply:
I took a look into our AutoML samples and found that it uses Evaluate() method after selecting the best model. In user's sample however, CrossValidate() is used instead of Evaluate(). Then I realized it's likely that user didn't code correctly since CrossValidate() is a training process. The data is trained twice that's why Fit() been called twice.

We might want to change the name of CrossValidate() to CrossValidateTraining()? As I think it's sort of confusing and let people easily link this with inference process.